pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -376,15 +411,57 @@ func ValidateServiceaccountMapping(p *v1beta1.PipelineSpec, pr *v1beta1.Pipeline
 // It will retrieve the Resources needed for the TaskRun using the mapping of providedResources.
 func ResolvePipelineRun(
 	ctx context.Context,
+	pipelineSpec v1beta1.PipelineSpec,
 	pipelineRun v1beta1.PipelineRun,
 	getTask resources.GetTask,
 	getTaskRun resources.GetTaskRun,
 	getClusterTask resources.GetClusterTask,
 	getCondition GetCondition,
+	providedResources map[string]*resourcev1alpha1.PipelineResource,
+) (*ResolvedPipelineRun, error) {
+
+	// We always validate the pipeline spec before resolving it, so errors are not
+	// expected here.
+	d, err := dag.Build(v1beta1.PipelineTaskList(pipelineSpec.Tasks))
+	if err != nil {
+		return nil, err
+	}
+	final, err := dag.Build(v1beta1.PipelineTaskList(pipelineSpec.Finally))
+	if err != nil {
+		return nil, err
+	}
+
+	state, err := resolvePipelineRun(ctx, pipelineSpec.Tasks, pipelineRun, getTask, getTaskRun, getClusterTask,
+		getCondition, providedResources)
+	if err != nil {
+		return nil, err
+	}
+
+	finalState, err := resolvePipelineRun(ctx, pipelineSpec.Finally, pipelineRun, getTask, getTaskRun, getClusterTask,
+		getCondition, providedResources)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ResolvedPipelineRun{
+		Dag:        d,
+		Final:      final,
+		DagState:   state,
+		FinalState: finalState,
+		WontRun:    skippedTasks(state, d),
+	}, nil
+}
+
+func resolvePipelineRun(
```

Comment:
We had this similar function implemented in first version of finally [PR](https://github.com/tektoncd/pipeline/pull/2437/files#diff-4e301064fef70f3c0db810267c0959f2R367).

This was simplified by merging both states (pipelineRunState) into one.

Reply:
My impression after working on this refactor is that overall it's much simpler like this. 

Having the dag handled in `pipelinerunresolution` allows us to have the best of both worlds:
- we can expose a single state to the reconciler, via the `State()` function
- we can work internally with two separated states, which makes a lot of the functions more natural, it removes the need to check which state is in which dag which we had to do in many places. 

The unit test coverage still need some work, as it might be best to target tests to `resolvePipelineRun` and only add a couple of tests for logic in `ResolvePipelineRun`.

Reply:
I understand @afrittoli the supporting functions are more natural this way but how will we address a scenario with a `finally` task having an input `PipelineResource` which refers to the output of a `DAG` task when resolved DAG tasks are stored in `DAGState` and resolved `finally` tasks are in `FinalState`. And same for task results.

Right now, `DAG` tasks and `finally` tasks are independent and have no reference from one to the other but at some point we need to address [these](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#known-limitations) limitations.

Reply:
e.g.

```
func convertParamsToResultRefs(pipelineRunState PipelineRunState, target *ResolvedPipelineRunTask) (ResolvedResultRefs, error) {
```
and 

```
func resolveResultRef(pipelineState PipelineRunState, resultRef *v1beta1.ResultRef) (*ResolvedResultRef, error) {
```

Reply:
I see, you have `State()` [here](https://github.com/tektoncd/pipeline/pull/2821/files#diff-0e0c01fdbdc46d2b6054a706f99516d0R482)

Reply:
:)

Besides, I feel that the handling of results as it's implemented today kind of breaks the abstraction of resolved pipeline tasks and runs. It might be good to move the result resolution into the resources module too.
The fact that we need to reconcile pipeline results in the `isDone` after completion time is already set, is an example of the issue. When we start sending cloud events for pipeline run events, this will cause the pipeline results to be missing in the pipeline run spec sent in the event.

Reply:
Yup I agree and we have the same issue with task results (task results are set after task run is declared successful, see PR #2538).

Moving the result resolution into the resources module could be next in refactoring queue.